### PR TITLE
Dispose HttpClient properly in HttpConnection

### PR DESCRIPTION
**Pull request by BrennanConroy**
#860 
Added tests for both scenarios @moozzyk called out in the linked issue.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs [hunk 1]
```diff
@@ -365,6 +364,10 @@ private void SetTransferMode(TransferMode transferMode)
                 Output.TryComplete(ex);
                 _logger.ErrorReceiving(_connectionId, ex);
             }
+            finally
+            {
+                _httpClient.Dispose();
```
Comment (moozzyk) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs]: How is the client disposed when the connection is never started? I think it won't be...

Comment (BrennanConroy) [reply src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs]: Interesting point, I don't think it ever was before?

Comment (moozzyk) [reply src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs]: It did - this is why we had it in the `DisposeAsync`

Comment (BrennanConroy) [reply src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs]: Why would you call `DisposeAsync` when the connection isn't started

Comment (moozzyk) [reply src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs]: I don't think it matters (maybe the app has some additional logic between creating and starting the connection to decide) - we should dispose properly when the user called `DisposeAsync()`...

Comment (davidfowl) [reply src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs]: So now we call Dispose in 3 places!? Can you write a comment saying why?

File: src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs [hunk 2]
```diff
@@ -412,7 +419,9 @@ private void SetTransferMode(TransferMode transferMode)
 
             if (Interlocked.Exchange(ref _connectionState, ConnectionState.Disconnected) == ConnectionState.Initial)
             {
-                // the connection was never started so there is nothing to clean up
+                // The connection was never started.
+                // Connection is now unusable, so cleanup resources.
+                _httpClient.Dispose();
```
Comment (moozzyk) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs]: I think the client will not be disposed if DisposeAsync is called when the connection is in the connecting state - if this [condition](https://github.com/aspnet/SignalR/pull/999/files#diff-ffe775cce0a6518263da4dfc756b867eR157) is not true.

**Comment by moozzyk** [comment]
I propose to postpone this change for now until we decide whether we want to make the connection instance reusable. If we will make the connection instance reusable (the connection is restartable) we will have to change how we dispose the HttpClient - it probably will be much easier.

**Comment by BrennanConroy** [comment]
Postponing

**Pull request closed without merging**